### PR TITLE
Make `std::os::fd` public.

### PR DESCRIPTION
**Pull request by sunfishcode**
`std::os::fd` defines types like `OwnedFd` and `RawFd` and is common
between Unix and non-Unix platforms that share a basic file-descriptor
concept. Rust currently uses this internally to simplify its own code,
but it would be useful for external users in the same way, so make it
public.

This means that `OwnedFd` etc. will all appear in three places, for
example on unix platforms:
 - `std::os::fd::OwnedFd`
 - `std::os::unix::io::OwnedFd`
 - `std::os::unix::prelude::OwnedFd`

r? @joshtriplett

### CODE REVIEW

**Comment by dtolnay** [comment]
Would it be possible to get a listing of everything that is being proposed to be stabilized at a path inside `std::os::fd` (and on what platforms) by this PR? Currently the summary just says "types like OwnedFd and RawFd" which isn't specific enough for me to approve the stabilization.

@rfcbot concern what are the contents

**Comment by sunfishcode** [comment]
@dtolnay The contents of the `fd` module in this PR are:
 - types: `OwnedFd`, `BorrowedFd`, `RawFd`
 - traits: `AsFd`, `AsRawFd`, `IntoRawFd`, `FromRawFd`

The platforms this PR adds the `fd` module to are Unix-family platforms (`cfg(unix)`) and WASI (`cfg(target_os = "wasi")`).

In the future, it's possible that fortanix_sgx and/or solid may want to add this module as well, as they also have similar types and traits.

**Comment by dtolnay** [comment]
Great!

@rfcbot resolve what are the contents
@rfcbot reviewed

**Comment by BurntSushi** [comment]
Is this insta-stable?

The other thing I'm confused about here is why things like `RawFd` are going in a platform independent module? They aren't used on Windows, for example, right?

**Comment by dtolnay** [comment]
I think the idea is that we already have `std::os::unix`, and `std::os::wasi`, and in the future perhaps `std::os::fortanix_sgx` or `std::os::solid`, all of which contain identical fd-related wrappers and traits. If someone is writing code for "anything with an fd", they can't easily do it. So `std::os::fd` serves as `std::os::[anything with an fd]`.

**Comment by BurntSushi** [comment]
@dtolnay Ah I see thanks, I think that makes sense.

@sunfishcode Can we add that guidance to the docs of `std::os::fd`? That is, call out that `RawFd` (and company) aren't purely platform independent, but that they are used on so many platforms that it can be convenient to just import it from `std::os::fd`.

I do admit though that this feels a little strange to me to be honest. I like the practicality behind this, but I think you still have to gate `std::os::fd` imports with conditional compilation in order for it to be correct, right? If so, I'd expect that to be in the docs too, because the import doesn't really hint that you need any conditional compilation.

**Comment by sunfishcode** [comment]
> Is this insta-stable?

Yes, I guess it currently is. Would it be better here to use `#[unstable(...)]`? And I guess open a tracking issue?

> @sunfishcode Can we add that guidance to the docs of `std::os::fd`? That is, call out that `RawFd` (and company) aren't purely platform independent, but that they are used on so many platforms that it can be convenient to just import it from `std::os::fd`.

I've clarified the comment to explicitly mention both Unix and WASI.

> I do admit though that this feels a little strange to me to be honest. I like the practicality behind this, but I think you still have to gate `std::os::fd` imports with conditional compilation in order for it to be correct, right? If so, I'd expect that to be in the docs too, because the import doesn't really hint that you need any conditional compilation.

It does already have conditional compilation; the `pub mod fd` is gated on `#[cfg(any(unix, target_os = "wasi", doc))]`. Is that what you're referring to?

**Comment by BurntSushi** [comment]
> It does already have conditional compilation; the `pub mod fd` is gated on `#[cfg(any(unix, target_os = "wasi", doc))]`. Is that what you're referring to?

Hmmm, I suppose, yes. I was more referring to consumers of this API. The import path doesn't really make it clear (like other import paths do) that conditional compilation is needed. It's just a departure from what we've done, I think.

The fact that `pub mod fd` is gated on `any(unix, target_os = "wasi", doc)` does make it better. But still overall feels a little weird to me.

**Comment by BurntSushi** [comment]
> Yes, I guess it currently is. Would it be better here to use `#[unstable(...)]`? And I guess open a tracking issue?

I'm not sure of the underlying mechanism, but basically, if we don't have to do this as insta-stable, then we probably shouldn't. The typical insta-stable things are trait impls, and I believe that's because there just isn't any alternative.

**Comment by sunfishcode** [comment]
> > Yes, I guess it currently is. Would it be better here to use `#[unstable(...)]`? And I guess open a tracking issue?
> 
> I'm not sure of the underlying mechanism, but basically, if we don't have to do this as insta-stable, then we probably shouldn't. The typical insta-stable things are trait impls, and I believe that's because there just isn't any alternative.

I've now filed a tracking issue https://github.com/rust-lang/rust/issues/98699, and updated the PR to use a `#[unstable(...)]` attribute. This was a little tricky, because the `fd` module is already used internally in a way that doesn't want it to be `#[unstable(...)]`, so it appears this requires having two separate modules.

**Comment by dtolnay** [comment]
@rfcbot fcp cancel

File: library/std/src/os/fd/raw.rs [hunk 1]
```diff
@@ -40,10 +40,7 @@ pub trait AsRawFd {
     /// ```no_run
     /// use std::fs::File;
     /// # use std::io;
-    /// #[cfg(unix)]
-    /// use std::os::unix::io::{AsRawFd, RawFd};
-    /// #[cfg(target_os = "wasi")]
-    /// use std::os::wasi::io::{AsRawFd, RawFd};
+    /// use std::os::fd::{AsRawFd, RawFd};
```
Comment (m-ou-se) [inline library/std/src/os/fd/raw.rs:43]: This shouldn't compile, if the stability attributes work correctly.

**Comment by Mark-Simulacrum** [comment]
I believe type aliases should allow for different stability attributes (recently used by the move of CStr to core and CString to liballoc).

**Comment by sunfishcode** [comment]
Applying the attributes to the module is what the current PR does.

I can see how to use type aliases for the types, but I don't see a way to do the same thing for traits.

File: library/std/src/os/fd/raw.rs (hunk 1 above)
Comment (sunfishcode) [reply library/std/src/os/fd/raw.rs:45]: It does compile, somehow. The doc tests pass. I don't understand stability attributes enough to know what's going on.

**Comment by sunfishcode** [comment]
I've now rebased this on https://github.com/rust-lang/rust/pull/99723, which simplifies it.

I am not sure whether the stability attributes are working as expected though.

**Comment by joshtriplett** [comment]
Since this isn't insta-stable anymore:

@bors r+

**Comment by Dylan-DPC** [comment]
@bors r- 

failed in [rollup](https://github.com/rust-lang/rust/pull/101282#issuecomment-1234513560)

**Comment by sunfishcode** [comment]
Fixed the WIndows build issue.

r? @joshtriplett

File: library/std/src/os/fd/mod.rs [hunk 2]
```diff
@@ -1,16 +1,25 @@
 //! Owned and borrowed Unix-like file descriptors.
+//!
+//! This module is supported on Unix platforms and WASI, which both use a
+//! similar file descriptor system for referencing OS resources.
 
 #![stable(feature = "io_safety", since = "1.63.0")]
```
Comment (bjorn3) [inline library/std/src/os/fd/mod.rs:6]: This should be `1.66.0`, right?

Comment (sunfishcode) [reply library/std/src/os/fd/mod.rs:6]: Good catch; I've now filed https://github.com/rust-lang/rust/pull/106867 to fix this.

File: library/std/src/os/wasi/io/mod.rs [hunk 3]
```diff
@@ -1,12 +1,6 @@
 //! WASI-specific extensions to general I/O primitives.
 
-#![deny(unsafe_op_in_unsafe_fn)]
 #![unstable(feature = "wasi_ext", issue = "71213")]
 
-mod fd;
-mod raw;
```
Comment (RalfJung) [inline library/std/src/os/wasi/io/mod.rs:7]: This removed two `mod` statements but the corresponding files still exist. I think they are dead now? Or is there some magic going on that somehow loads them anyway?

Comment (sunfishcode) [reply library/std/src/os/wasi/io/mod.rs:7]: They're indeed dead now. I've now filed #135512 to  remove them.

Comment (sunfishcode) [reply library/std/src/os/wasi/io/mod.rs:7]: And I've now noticed that https://github.com/rust-lang/rust/pull/135491 already does this, so closing in favor off that.

**Pull request merged**